File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.js
/**
 * @param {string} s
 * @return {number}
 */
var minCut = function(s) {
  const inbound = i => {
    return i >= 0 && i < s.length;
  };
  
  const adj = {};
  for (let i = 0; i < s.length; i++) {
    for (let j = i; j < s.length && j <= i + 1; j++) {
      let [l, r] = [i, j];
      while (inbound(l) && inbound(r) && s[l] === s[r]) {
        adj[l] ||= new Set();
        adj[l].add(r + 1);
        l--;
        r++;
      }
    }
  }
  
  const pathQ = [[0]];
  const visited = new Set();
  visited.add(0);
  let currPath, currNode;
  while (pathQ.length) {
    currPath = pathQ.shift();
    currNode = currPath.at(-1);
    if (currNode === s.length) return currPath.length - 2;
    adj[currNode].forEach(nb => {
      if (!visited.has(nb)) {
        visited.add(nb);
        pathQ.push([...currPath, nb]);
      }
    })
  }
};

File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.rb
# @param {String} s
# @return {Integer}
def min_cut(s)
  adj = Hash.new{ |h, k| h[k] = Set.new }
  (0...s.length).each do |i|
    (i..i+1).each do |j|
      l, r = i, j
      while (0...s.length).include?(l) && (0...s.length).include?(r) && s[l] == s[r]
        adj[l] << r+1
        l, r = l-1, r+1
      end
    end
  end
  
  path_q, visited = [[0]], [0].to_set
  while !path_q.empty?
    curr_path = path_q.shift
    curr_node = curr_path[-1]
    return curr_path.length-2 if curr_node == s.length
    adj[curr_node].each do |nb|
      if !visited.include?(nb)
        visited << nb
        path_q << curr_path+[nb]
      end
    end
  end
end